### ISSUE TITLE
Spatialmaterial with own shader can not be local to scene

### ISSUE THREAD

**Issue by nerdnils** [godotengine/godot#52429]
### Godot version

3.3.3

### System information

Any

### Issue description

When writing/creating a spatial(must be spatial! issue does not exist in 2d with canvas shader) shader and applying the material using it to a mesh which is in a scene which is instanced, the material can not be "local to scene".

### Steps to reproduce

Create own (spatial! issue only exists in 3d! works fine with 2d shaders)shader, create a material which uses it. Make material "lcoal to scene".

Now create a scene with a mesh using that meratial. Instance that scene, change any uniform of the material.

Expected behaviour: Only one mesh changes that uniform(for example color) since the material is checked to be local to this scene only.

What really happens: All materials change.

Expected behaviour comes up when doing this with a canvas shader in 2d. But in 3d this bug comes up.

### Minimal reproduction project

Can not at the moment I am sorry but I had people in the chat and found a forum thread with people having the same issue since february. Bug still exists in newest version.

**Comment by nisovin** [comment]
Since I was looking into this in discord, I'll provide the test scenes I was using.

[BugShaderMaterial.zip](https://github.com/godotengine/godot/files/7117495/BugShaderMaterial.zip)

I tried putting the material in every property it can go in, but it always had the same result. I tested in 3.4 beta and 3.3.2. It works fine in 2D, but not 3D. Uncommenting the duplicate() call allows it to work properly.

**Comment by clayjohn** [comment]
@nisovin In your MRP "local to scene" is not checked on the material that the cube uses, so all instanced cubes share a reference to the same material. However "local to scene" is checked on the material used by the Sprite, so each sprite instance has its own material. 

Checking "local to scene" on the cube's material makes it work as expected.

**Comment by nisovin** [comment]
@clayjohn It is checked though... have I not checked the correct thing? As far as I can tell, it is checked, but it doesn't work.
![20210907_183229_1631061149_wVU1r](https://user-images.githubusercontent.com/464568/132426900-41af56e5-8762-45e4-aef6-f4c0161e3f9d.png)

**Comment by clayjohn** [comment]
@nisovin You have checked it on the shader itself, but not on the material. If you scroll down a little you will see the field for the material.

**Comment by nisovin** [comment]
I realized that right as you replied. How embarrassing...

**Comment by clayjohn** [comment]
@nerdnils can you confirm that your issue is the same as nisovin's and can be fixed by setting the "local to scene" flag on the material?

**Comment by nerdnils** [comment]
I am sorry but I need a few days before I can test/veryify. Will post here after I tested. Thanks for some pation the seemingly(can't reach my code now) solution.

**Comment by nerdnils** [comment]
https://pasteall.org/media/f/2/f28e377148f97684daf2fec628fc7e76.png

You mean the one on the lower right here? Was already checked for me before I even mentioned the issue on discord. Will copy and slaughter my project to get a minimal demonstration project.

**Comment by nerdnils** [comment]
Here is my file: https://ufile.io/mjo9nr7y

Change the furcolor of any mouse, the furcolor of the other one will change, too.
Mesh, Material, Shader, anything is local to scene.

**Comment by nisovin** [comment]
@nerdnils This is likely because you've just duplicated the nodes. If you create scenes and instance them individually, it should work.

**Comment by nerdnils** [comment]
> @nerdnils This is likely because you've just duplicated the nodes. If you create scenes and instance them individually, it should work.

In the original it is a scene. And i instance them individually in script:
header:
onready var mousecharacter_scene = preload("res://player/mouse.tscn")
later when a player joins:
character = mousecharacter_scene.instance()

and then
quadnode.get_active_material(0).albedo_color=playersettings[id].get_textcolor()
changes the colour of ALL mice.

**Comment by nerdnils** [comment]
Should I create another minimal project to prove it?

**Comment by clayjohn** [comment]
@nerdnils yes please. I ran your other MRP quickly, but it didn't have multiple instances of the scene and I didn't have a chance to create an MRP of my own to test it out.

**Comment by nerdnils** [comment]
> @nerdnils yes please. I ran your other MRP quickly, but it didn't have multiple instances of the scene and I didn't have a chance to create an MRP of my own to test it out.

While creating the screne I noticed it works if the mesh is the root node of that scene. It does not work if the mesh is deeper in the other scene. But it has to when I use it as a character for a player because Kinematicbody, Armature etc has to be above it. In the editor I can not even find a way to directly change the value of the furcolor in the editor(I can only do it in the editor if the mesh having the material is the root node of the other scene). I can do it through script but this invokes the error.

So this solves why it works in the project of nisovin but not when using the material for anything animated or used in the physics engine ingame.

When making the mesh the rootnode, adapting the colors and changing the order in the imported scene afterwards, the colors also change to all be the same without changing anything else.

I can not make the meshnode the root node. So I need a solution to make it work when it isn't. And please don't tell me the solution is to create all collision nodes, armatures, kinematicbodys and what not in the scene i want to import to from scratch every single time.

And of course: here you are: https://ufile.io/axo7af86 :) change node order in res://characters/mouse/mousemenu.tscn to test what I found out.

**Comment by nerdnils** [comment]
Can you now see what the problem is? It should also work with copied objects imho. Materials need a "local to object/mesh"-button if you ask me.

**Comment by clayjohn** [comment]
> Can you now see what the problem is? It should also work with copied objects imho. 

Taking a look now

> Materials need a "local to object/mesh"-button if you ask me.

They do, it is called "make unique"
